feat: implement server-side contribution streaks and longest streak tracking

vivekyadav-3 opened the pull request:
## 🎯 Overview
This PR adds server-side contribution streak calculation and introduces tracking for the longest streak on contributor profiles.
## ✨ What’s Included
### Backend
* Moved streak calculation from the client to the server API.
* Added a `calculateStreaks` helper that determines:
  * **Current Streak**: Consecutive active days leading up to today or yesterday.
  * **Longest Streak**: The contributor’s highest consecutive active-day record.
* This improves consistency, speeds up client rendering, and makes the data reusable across the app.
### Frontend
* Updated the UI to use `current_streak` and `longest_streak` directly from the API.
* Added a visual element for the **Longest Streak** (🏆).
* Added a subtle pulse animation to highlight the current streak.
* Removed old client-side streak logic to keep things clean.
## 🛠️ Technical Notes
* **Time complexity**: O(n), where n is the number of active days.
* **Type safety**: Uses the existing `DailyActivity` and `ContributorEntry` interfaces.
* **Edge cases**: Added safeguards for sorting activity dates and handling empty activity lists.
## ✅ Testing
* [x] Current streak logic (today/yesterday flow) works correctly
* [x] Longest streak updates properly
* [x] UI animation works as expected
* [x] Lint checks passed
* [x] Mobile layout verified

## 🔗 References

* Fixes #49
* Synced with discussion on Slack with @Naman-Chhabra

cc: @Naman-Chhabra @Aboobacker-MK @Supreeth-C


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

* **New Features**
  * Added streak tracking for contributors, displaying current and longest consecutive contribution streaks
  * Introduced an achievement badges system recognizing contributors based on activity levels and roles
  * Added activity visualization chart breaking down contribution types by count and impact
  * Enhanced contributor profiles with additional performance metrics including PR turnaround and activity insights
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ShinichiShi commented:
Hey @vivekyadav-3 great work, just a few suggestions:
- remove the markdown files as they are unnecessary 
- in the collaborator card I can see 🔥/🏆, i didn't get what 🔥 indicates here ?
- fix the coderabbit mentioned issues

Also I'm only supreeth, tag me as @ShinichiShi 😄

vivekyadav-3 commented:
Hey @naman79820, thanks for calling those out! I pushed a refactor to cover
everything you mentioned:

   -

   *Streak logic* is fixed now. It properly checks consecutive days with
   actual date math instead of just counting active days.
   -

   *Code duplication* is gone. I moved the streak calculation into a shared
   utility (lib/streak-utils.ts).
   -

   *Types* are cleaned up and replaced with proper interfaces instead of any
   .
   -

   *History cleanup* is done too. Removed the personal/draft files so the
   PR only focuses on the feature.

Should be good for another review



On Thu, Jan 1, 2026 at 12:31 AM Naman Chhabra ***@***.***>
wrote:

> *naman79820* left a comment (CircuitVerse/community-dashboard#55)
> <https://github.com/CircuitVerse/community-dashboard/pull/55#issuecomment-3702723554>
>
> Hey @vivekyadav-3 <https://github.com/vivekyadav-3> can you please
> resolve coderabbit and copilot suggestions please :))
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/CircuitVerse/community-dashboard/pull/55#issuecomment-3702723554>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BSGQEXGZ72XDZNTT6JOGXIT4EQMQRAVCNFSM6AAAAACQL4VAAKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTOMBSG4ZDGNJVGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

-- 


* <http://www.kiit.ac.in>*

*
Go Green: Kindly don't print this unless so 
required.*

*
*

*Established U/S 3 of UGC Act and Accredited by NBA of 
AICTE and NAAC of UGC*



Visit us @ *http://www.kiit.ac.in 
<http://www.kiit.ac.in/>*

Follow us @* **|***Twitter 
<https://www.twitter.com/KIITUniversity/>***|***Facebook** 
<https://www.facebook.com/KIITUniversity>*|***Instagram 
<https://www.instagram.com/KIITUniversity/>***|*


 
  
  

The information 
contained in this electronic message and any
  attachments to this message 
are intended for the exclusive use of the
  addressee(s) and may contain 
proprietary, confidential or privileged
  information. If you are not the 
intended recipient, you should not
  disseminate, distribute or copy this 
e-mail. Please notify the sender
  immediately and delete all copies of 
this message and any attachments.
  

 VIRUS WARNING: Computer viruses can 
be
  transmitted via email. The recipient should check this email and any
  
attachments for the presence of viruses. The institute accepts no liability
  for any damage caused by any virus transmitted by this email.

naman79820 commented on lib/db.ts:
```diff
@@ -141,45 +141,67 @@ export async function getRecentActivitiesGroupedByType(
-export async function getContributor(_username: string) {
-  return null;
-}
+  if (!contributor) return { contributor: null, activities: [], totalPoints: 0, activityByDate: {} };
```
@vivekyadav-3

naman79820 commented on lib/streak-utils.ts:
```diff
@@ -0,0 +1,89 @@
+          tempStreak++;
+        } else {
+          tempStreak = 1;
+        }
```
@vivekyadav-3

naman79820 commented:
i think 2 of the suggestions done by coderrabit are still pending can you have a look please :)) @vivekyadav-3

vivekyadav-3 commented:
Done! I’ve added the defensive checks and unified the rounding logic as
suggested. Please take another look when you get a chance.



On Fri, Jan 2, 2026 at 3:04 PM Naman Chhabra ***@***.***>
wrote:

> *naman79820* left a comment (CircuitVerse/community-dashboard#55)
> <https://github.com/CircuitVerse/community-dashboard/pull/55#issuecomment-3704865320>
>
> i think 2 of the suggestions done by coderrabit are still pending can you
> have a look please :)) @vivekyadav-3 <https://github.com/vivekyadav-3>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/CircuitVerse/community-dashboard/pull/55#issuecomment-3704865320>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BSGQEXCRGR2LDKMOK7KHWQT4EY3RVAVCNFSM6AAAAACQL4VAAKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTOMBUHA3DKMZSGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

-- 


* <http://www.kiit.ac.in>*

*
Go Green: Kindly don't print this unless so 
required.*

*
*

*Established U/S 3 of UGC Act and Accredited by NBA of 
AICTE and NAAC of UGC*



Visit us @ *http://www.kiit.ac.in 
<http://www.kiit.ac.in/>*

Follow us @* **|***Twitter 
<https://www.twitter.com/KIITUniversity/>***|***Facebook** 
<https://www.facebook.com/KIITUniversity>*|***Instagram 
<https://www.instagram.com/KIITUniversity/>***|*


 
  
  

The information 
contained in this electronic message and any
  attachments to this message 
are intended for the exclusive use of the
  addressee(s) and may contain 
proprietary, confidential or privileged
  information. If you are not the 
intended recipient, you should not
  disseminate, distribute or copy this 
e-mail. Please notify the sender
  immediately and delete all copies of 
this message and any attachments.
  

 VIRUS WARNING: Computer viruses can 
be
  transmitted via email. The recipient should check this email and any
  
attachments for the presence of viruses. The institute accepts no liability
  for any damage caused by any virus transmitted by this email.

naman79820 commented:
Two more suggestions have come in from coderabbitai @vivekyadav-3 
Please fix all of those and give it some time, if no new suggestions come back then tag me or Atharva for review.
Thanks! 😊

naman79820 commented:
All the checks are failing now

vivekyadav-3 commented:
Fixed! I've pushed the missing badge system files (AchievementBadges.tsx and badges.ts). The build should pass now. Sorry for the oversight!

vivekyadav-3 commented:
Hi Naman,

Everything’s sorted now:

   -

   Missing badge files have been added
   -

   Merge conflicts are resolved
   -

   CodeRabbit suggestions are implemented
   -

   The branch merges cleanly with main

The PR is ready for a final review. Sorry again for the earlier confusion
around the missing files. I’ve cleaned up my git workflow to avoid this
going forward.

Thanks for your patience!



On Fri, Jan 2, 2026 at 7:15 PM Naman Chhabra ***@***.***>
wrote:

> *naman79820* left a comment (CircuitVerse/community-dashboard#55)
> <https://github.com/CircuitVerse/community-dashboard/pull/55#issuecomment-3705353500>
>
> All the checks are failing now
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/CircuitVerse/community-dashboard/pull/55#issuecomment-3705353500>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BSGQEXFDJBJNEO3Z4HYCQ6T4EZY6TAVCNFSM6AAAAACQL4VAAKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTOMBVGM2TGNJQGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

-- 


* <http://www.kiit.ac.in>*

*
Go Green: Kindly don't print this unless so 
required.*

*
*

*Established U/S 3 of UGC Act and Accredited by NBA of 
AICTE and NAAC of UGC*



Visit us @ *http://www.kiit.ac.in 
<http://www.kiit.ac.in/>*

Follow us @* **|***Twitter 
<https://www.twitter.com/KIITUniversity/>***|***Facebook** 
<https://www.facebook.com/KIITUniversity>*|***Instagram 
<https://www.instagram.com/KIITUniversity/>***|*


 
  
  

The information 
contained in this electronic message and any
  attachments to this message 
are intended for the exclusive use of the
  addressee(s) and may contain 
proprietary, confidential or privileged
  information. If you are not the 
intended recipient, you should not
  disseminate, distribute or copy this 
e-mail. Please notify the sender
  immediately and delete all copies of 
this message and any attachments.
  

 VIRUS WARNING: Computer viruses can 
be
  transmitted via email. The recipient should check this email and any
  
attachments for the presence of viruses. The institute accepts no liability
  for any damage caused by any virus transmitted by this email.
